### PR TITLE
Fix for issue actionType is not checked before calling searchContextProviders() in updateEntity() function #2874

### PR DESCRIPTION
**Pull request by SourabhChourasiya-NEC**
Fix for issue **actionType is not checked before calling searchContextProviders() in updateEntity() function #2874**  [https://github.com/telefonicaid/fiware-orion/issues/2874]

### CODE REVIEW

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 1]
```diff
@@ -3416,8 +3416,10 @@ static void updateEntity
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: An entry in CHANGES_NEXT_RELEASE file regarding this fix should be added, e.g:

```
- Fix: check for providers is done only in UPDATE or REPLACE actionType case in update entity logic (#2874)
```

**Comment by fgalan** [comment]
Thanks for the contribution!

The PR is fine in general (appart from a minor comment regarding CHANGES_NEXT_RELEASE file). However, lookint to the parent issue:

> We are not sure if we have a functional test covering update context APPEND with CPr registration in place covering this case. That should be the first thing to check.

Have you checked that? In the case there isn't any .test covering the case, would it be possible to add a new one for it? Or it is too complex?

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
I have checked and there is no functional test case present for update context APPEND with CPr registration. As you told, i will prepare a new functional test case for the same and let you know.
Thank you

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,

I have wrriten the test cases as follows:

1. registring a context provider for entity Room, identify by ID ConferenceRoom & OfficeRoom, and with attribute     
   temperature to context Broker.
   
(curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/json' \
   --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
  "contextRegistrations": [
    {
      "entities": [
        {
          "type": "Room",
          "isPattern": "false",
          "id": "ConferenceRoom"
        },
        {
          "type": "Room",
          "isPattern": "false",
          "id": "OfficeRoom"
        }
      ],
      "attributes": [
        {
          "name": "temperature",
          "type": "degree",
          "isDomain": "false"
        }
      ],
      "providingApplication": "http://localhost:1028/application"
    }
  ],
  "duration": "PT1H"
}
EOF

2. Now I am requesting a query of actionType APPEND for the same entity with ID OfficeRoom with same attribute.

(curl localhost:1026/v1/updateContext -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "contextElements": [
        {
            "type": "Room",
            "isPattern": "false",
            "id": "OfficeRoom",
            "attributes": [
                {
                    "name": "temperature",
					"type": "degree",
					"value": "12"
     			}
            ]
        }
    ],
    "updateAction": "APPEND"
} 
EOF

It gives me a response code 200, but the problem is that i am not able to identify that what type of operation was performed by this query, either creating an entity locally or updating an existing entity.
The response is as follows:

"statusCode": {
                "code": "200",
                "reasonPhrase": "OK"
            }

However to confirm the type of operation i used GDB. By attaching a debugger i have identified that there is no call made to searchContextProviders() method from updateEntity() method.

(gdb) b updateEntity
Breakpoint 1 at 0x65a7d0
(gdb) b searchContextProviders
Breakpoint 2 at 0x648ad0
(gdb) c
Continuing.
[New Thread 0x7f77e5050700 (LWP 4937)]
[Switching to Thread 0x7f77e5050700 (LWP 4937)]

Breakpoint 1, 0x000000000065a7d0 in updateEntity(mongo::BSONObj const&, std::string const&, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, std::string const&, ContextElement*, UpdateContextResponse*, bool*, std::string*, ApiVersion, std::string const&, std::string const&) ()
(gdb) c
Continuing.
[Thread 0x7f77e5050700 (LWP 4937) exited]

Please guide me how can i identify this from query response
Thank you

**Comment by fgalan** [comment]
You could use the following script for the .test

Start two CBs: a "real" CB listening in the usual port and a CB acting as CP in a different port. Have a look for instance to query_redirect_operation.test to see an example (SHELL-INIT section).

Then the steps whould be:

1. Register CP
2. Create an entity in CB with POST /v2/op/update (better than the NGSIv1 variant you show in your example)
3. Query to CP: see nothing (that's ensures the APPEND wasn't forwarded)
4. Query to CB: see the entity (it was created locally)

Makes sense?

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
Thanks for your continuous support. I have prepared a functional test case according to following condition: 
1. Register CP
2. Create an entity in CB with POST /v2/op/update (better than the NGSIv1 variant you show in your example)
3. Query to CP: see nothing (that's ensures the APPEND wasn't forwarded)
4. Query to CB: see the entity (it was created locally)

Results have been added in already uploaded file “updateContext_APPEND_CPr_Reg.test”.
Please verify it, and guide me for further action.
Thank you

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        }
+      ],
+      "attributes": [
+        {
+          "name": "temperature",
+          "type": "degree",
+          "isDomain": "false"
+        }
+      ],
+      "providingApplication": "${Context_Provider_URL}"
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Where the `context_Provicer_URL` variable is defined?

Please, look how this is done in other .test using CB as CP to set correctly the `providingApplication`.

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    }
+  ],
+  "duration": "PT1H"
+}'
+orionCurl --url "/v1/registry/registerContext" --payload "$payload"
+REG_ID=$(echo "$_response" | grep registrationId | awk -F\" '{ print $4}')
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: This REG_ID is not used in next steps in this .test, so this line could be removed.

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+payload='{
+    "contextElements": [
+        {
+            "type": "Floor",
+            "isPattern": "false",
+            "id": "Floor1",
+            "attributes": [
+                {
+                    "name": "temperature",
+					"type": "degree",
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: This line doesn't seem to be correctly indented. Probably some tab were used. Please, don't use tabs for indent, only spaces.

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      ],
+      "providingApplication": "${Context_Provider_URL}"
+    }
+  ],
+  "duration": "PT1H"
+}'
+orionCurl --url "/v1/registry/registerContext" --payload "$payload"
+REG_ID=$(echo "$_response" | grep registrationId | awk -F\" '{ print $4}')
+
+#
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: L62-L64 could be removed. The "echo 02.xxxx" is enough.

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+# 02. Creating an entity in CB with actionType APPEND by using NGSIv1 .
+#
+echo "02. Creating an entity in CB with actionType APPEND by using NGSIv1."
+echo "========================================================================"
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Adjust line length to the one above. I mean, remove "=" in L66 so it gets same length that L65

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 7]
```diff
@@ -0,0 +1,212 @@
+# Copyright 2014 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: 2014 -> 2018

**Comment by fgalan** [comment]
I have had a look to the test and it looks good for being the first one you write :) Please have a look to the comments I've provided and fix them.

Once you have finished, please include in this PR the output of `make ft` command as comment in this PR (use http://gist.github.com if the output is too large) in order to check that all works as expected.

Thanks!

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 7 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Hi @fgalan Sir,
I have applied all the corrections on .test file, Please review it and guide me further.
Thank You !

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

**Comment by fgalan** [comment]
Changes seems to be fine, thanks! Howver, in order to follow the "PR contribution protocol" (described in detail here https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/contribution_guidelines.md#pull-request-protocol) please review all the comments (some of them are now hiden and you have to click on "show outdated") and add a `"Fixed in <commithash>"` comment in each one. I have done it in the first one, as example.

In addition, please add to this PR (as a general comment) what I requested in my previous comment:

> Once you have finished, please include in this PR the output of `make ft` command as comment in this PR (use http://gist.github.com if the output is too large) in order to check that all works as expected.

Thanks!

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 3 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 4 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 5 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 6 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 7 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in 589148e

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir

thanks for your continuous support.
As you know i have prepared .test case file and also verified it on two real CB's. But now when i am trying to verify it from "make ft", it gives me --SHELL-INIT-- ERROR 11


[root@fiware-env fiware-orion]# cat test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.shellInit.stdout
{ "ok" : 1 }
{ "ok" : 1 }
Unable to start contextBroker

Please guide how can i overcome from this issue.

Thank you

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan  Sir,

Thank you for your continuous support.
As you know I'm facing an issue related to functional test. 
For this I've performed an experiment, I've taken an existing .test file and replaced the content of that .test file from my .test file.
By doing so, fuctional test has been successfully executed. 
This may be a bug that why I'm not allowed to perform "make ft" on newly added .test file and also it gives me an error "unable to start contextBroker".

For now, could you please tell me how can I resolve this problem. I want to execute my newly created .test file to test this fix.

Thank you

**Comment by fgalan** [comment]
First, try to run your new test isolately. In order to to so, first check:

1. That contextBroker is in your path file (typically, using `contextBroker --version`, ensuring that the githash you see there is the one corresponding with your changes), then go to the test/functionalTest directory 
2. Your MongoDB instance runs correctly (i.e. `mongo localhost`).

Then, run:

```
./testHarness.sh <path to your .test file>
```

Does it work?

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir
thanks for your continuous support. I have successfully performed functional test and also uploaded make ft output on  http://gist.github.com . Please verfity ce0cfa609578399169be64b69e39eff8 and guide me further
Thank you

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    ],
+    "attributes": [
+        "temperature"
+    ]
+}'
+orionCurl --url /v1/queryContext --payload "$payload"
+echo
+echo
+
+
```
Comment (kzangeli) [inline test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Just one doubt ...
This test seems to test something that has been working for a very long time ...
Am I wrong?
Have we suddenly started to forward appends?

Comment (fgalan) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: I think you are right: this has worked that way time ago. However, it seems we haven't covered that case, so it's a good addition.

Actually, the PR fixes a very internal aspect of the code. I think there is no easy way to test it "directly", so it is fine if regression is still working (including the new .test).

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 9]
```diff
@@ -3416,8 +3416,10 @@ static void updateEntity
 
   /* To finish with this entity processing, search for CPrs in not found attributes and
    * add the corresponding ContextElementResponse to the global response */
-  searchContextProviders(tenant, servicePathV, *enP, ceP->contextAttributeVector, cerP);
-
+  if((strcasecmp(action.c_str(), "update") == 0) || (strcasecmp(action.c_str(), "replace") == 0))
```
[the remark this answers is not in the document]
Comment (SourabhChourasiya-NEC) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 603dcdc and 4d8aa61

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+updateContext actionType APPEND with Context Provider Registration
+
+--SHELL-INIT--
+dbInit CB
+dbInit CP1
+brokerStart CB
+brokerStart CP1
+
+--SHELL--
+
```
[the remark this answers is not in the document]
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in b0e10f1

File: test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test (hunk 8 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Fixed in b0e10f1

Comment (fgalan) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Actually, I think this is most "NTC" than "Fixed in"... but it's ok anyway ;)

Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3177_updateContext_APPEND_CPr_Registration/updateContext_APPEND_CPr_Reg.test]: Hi @fgalan Sir and @kzangeli  Sir,
Thanks for your continuous support. 
I want to know that anything is remaining from my side on this issue. If any please guide me further, so we can close this issue ASAP.
Thank you

**Pull request merged**